lib/ignored-paths.js

```diff
@@ -21,57 +23,97 @@ debug = debug("eslint:ignored-paths");
 //------------------------------------------------------------------------------
 
 var ESLINT_IGNORE_FILENAME = ".eslintignore";
+var DEFAULT_IGNORE_PATTERNS = [
+    "/node_modules/",
+    "/bower_components/"
+];
+var DEFAULT_OPTIONS = {
+    dotfiles: false,
+    cwd: process.cwd()
+};
 
 
 //------------------------------------------------------------------------------
 // Helpers
 //------------------------------------------------------------------------------
 
+var ignoreFileFinder;
+
 /**
- * Load and parse ignore patterns from the file at the given path
- * @param {string} filepath Path to the ignore file.
- * @returns {string[]} An array of ignore patterns or an empty array if no ignore file.
+ * Find an ignore file in the current directory or a parent directory.
+ * @param {stirng} cwd Current working directory
+ * @returns {string} Path of ignore file or an empty string.
  */
-function loadIgnoreFile(filepath) {
-    var ignorePatterns = [];
-
-    /**
-     * Check if string is not empty
-     * @param {string} line string to examine
-     * @returns {boolean} True is its not empty
-     * @private
-     */
-    function nonEmpty(line) {
-        return line.trim() !== "" && line[0] !== "#";
+function findIgnoreFile(cwd) {
+    cwd = cwd || DEFAULT_OPTIONS.cwd;
+    if (!ignoreFileFinder) {
+        ignoreFileFinder = new FileFinder(ESLINT_IGNORE_FILENAME, cwd);
     }
 
-    if (filepath) {
-        try {
-            ignorePatterns = fs.readFileSync(filepath, "utf8").split(/\r?\n/).filter(nonEmpty);
-        } catch (e) {
-            e.message = "Cannot read ignore file: " + filepath + "\nError: " + e.message;
-            throw e;
-        }
-    }
+    return ignoreFileFinder.findInDirectoryOrParents(cwd);
+}
 
-    return ["node_modules/**", "bower_components/**"].concat(ignorePatterns);
+/**
+ * Replace Windows with Unix style paths and remove ./ prefix
+ * @param {string} filepath Path to normalize
+ * @returns {string} Normalized filepath
+ */
+function normalizeFilepath(filepath) {
+    filepath = filepath.replace("\\", "/");
+    filepath = filepath.replace(/^\.\//, "");
+    return filepath;
 }
 
-var ignoreFileFinder;
+/**
+ * Remove a prefix from a filepath
+ * @param {string} filepath Path to remove the prefix from
+ * @param {string} prefix Prefix to remove from filepath
+ * @returns {string} Normalized filepath
+ */
+function removePrefixFromFilepath(filepath, prefix) {
```

Comment:
This function is used to ensure patterns are relative to the `base` directory. When an absolute path is specified as pattern, the `base` prefix will be removed and with that the pattern is converted to a relative pattern instead. Btw. the `fs.realPathSync` call is necessary to resolve symlinks.